implementations/rust/ockam/ockam_api/src/logs/setup.rs

```diff
@@ -0,0 +1,492 @@
+            provider_builder = provider_builder.with_span_processor(span_processor);
+        }
+        None => provider_builder = provider_builder.with_simple_exporter(exporter),
+    };
+    if let Some(config) = trace_config {
+        provider_builder = provider_builder.with_config(config);
+    }
+    let provider = provider_builder.build();
+    let tracer = provider.versioned_tracer(
+        "ockam",
+        Some(env!("CARGO_PKG_VERSION")),
+        Some(SCHEMA_URL),
+        None,
+    );
+    let _ = global::set_tracer_provider(provider.clone());
+    (tracer, provider)
+}
+
+fn make_resource(app_name: String) -> Resource {
+    let host_name = gethostname().to_string_lossy().to_string();
+    Resource::new(vec![
+        KeyValue::new(
+            opentelemetry_semantic_conventions::resource::SERVICE_NAME,
+            "ockam",
+        ),
+        KeyValue::new(
+            opentelemetry_semantic_conventions::resource::HOST_NAME,
+            host_name,
+        ),
+        KeyValue::new(APP_NAME.clone(), app_name),
+    ])
+}
+
+/// TODO: make sure that this is parsed as a proper URL
+fn get_tracing_endpoint() -> String {
+    match std::env::var("OTEL_EXPORTER_OTLP_ENDPOINT") {
+        Ok(endpoint) => {
+            info!("OTEL_EXPORTER_OTLP_ENDPOINT defined as {endpoint}");
+            endpoint
+        }
+        Err(_) => {
+            info!("using the default value for OTEL_EXPORTER_OTLP_ENDPOINT: http://localhost:4317");
+            "http://localhost:4317".to_string()
+        }
+    }
+}
+
+fn get_otlp_headers() -> MetadataMap {
+    match std::env::var("OTEL_EXPORTER_OTLP_HEADERS") {
+        Ok(headers) => {
+            match headers.split_once('=') {
+                // TODO: find a way to use a String as a key instead of a &'static str
+                Some((key, value)) => {
+                    match (
+                        MetadataKey::from_bytes(key.as_bytes()),
+                        MetadataValue::try_from(value.to_string()),
+                    ) {
+                        (Ok(key), Ok(value)) => {
+                            let mut map = MetadataMap::with_capacity(1);
+                            map.insert(key, value);
+                            map
+                        }
+                        _ => MetadataMap::default(),
+                    }
+                }
+                _ => MetadataMap::default(),
+            }
+        }
+        _ => MetadataMap::default(),
+    }
+}
+
+#[derive(Debug)]
+pub struct TracingGuard {
+    _worker_guard: Option<WorkerGuard>,
+    logger_provider: LoggerProvider,
+    tracer_provider: opentelemetry_sdk::trace::TracerProvider,
+}
+
+impl TracingGuard {
+    pub fn force_flush(&self) {
+        self.logger_provider.force_flush();
+        self.tracer_provider.force_flush();
+    }
+
+    pub fn shutdown(&self) {
+        self.force_flush();
+        global::shutdown_tracer_provider();
+        global::shutdown_logger_provider();
+    }
```

Comment:
I've seen `flush` is used when shutting down a foreground node. Should this be used explicitly at the end of any command execution?

Reply:
`flush` is otherwise done [via the `TracingGuard` at the end of `run`](https://github.com/build-trust/ockam/pull/7438/files#diff-f5a30aef85b96ee0c58c26d9c22690c484ea212154879d380cae8bd4b27ce018R455). I have also added an explicit `global::shutdown_tracer_provider();` at the end of that function but I don't think it is necessary. This is something I can revisit with my next PR.

Reply:
OK got it, I didn't see that piece of code, thanks.